Tests/NIOTests/TestUtils.swift

```diff
@@ -294,3 +295,76 @@ func assertSuccess<Value>(_ result: Result<Value, Error>, file: StaticString = #
 func assertFailure<Value>(_ result: Result<Value, Error>, file: StaticString = #file, line: UInt = #line) {
     guard case .failure = result else { return XCTFail("Expected result to be a failure", file: file, line: line) }
 }
+
+func forEachActiveChannelType<T>(file: StaticString = #file,
+                                 line: UInt = #line,
+                                 _ body: @escaping (Channel) throws -> T) throws -> [T] {
+    var eventLoopBag: [MultiThreadedEventLoopGroup] = []
+    let lock = Lock()
+    defer {
+        XCTAssertNoThrow(try lock.withLock {
+            try eventLoopBag.forEach {
+                try $0.syncShutdownGracefully()
+            }
+        })
+    }
+
+    func freshEventLoop() -> EventLoop {
+        let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+        lock.withLock {
+            eventLoopBag.append(group)
+        }
+        return group.next()
+    }
+
+    // TCP
+    let tcpChildChannel = freshEventLoop().makePromise(of: Channel.self)
+    let tcpAcceptedChannel = freshEventLoop().makePromise(of: Channel.self)
+    let tcpServerChannel = try assertNoThrowWithValue(ServerBootstrap(group: freshEventLoop())
```

Comment:
The server channel and the accepted channel here are identical, no?

Reply:
No, the server channel is the actual `ServerSocketChannel` (the one we call `accept` on) and the accepted channel is the `SocketChannel` (the one we accepted). Type-wise, the _child_ channel and the _accepted_ channel are the same (`SocketChannel`) so we could totally reduce to only one.

Reply:
`tcpAcceptedChannel` is fulfilled in the `serverChannelInitializer`, and `tcpServerChannel` is the result of the call to `bind()`. I’m pretty sure that if you did an object identity comparison here they’d be identical.

Put another way, I think `tcpAcceptedChannel` is actually of type `ServerSocketChannel`. I think the promise naming here has led to confusion. You can observe the result of this confusion at the bottom of this file, where your loop code is over `tcpServerChannel`, `tcpChildChannel`, `tcpAcceptedChannel`, and `udpChannel`. This does not pass the `tcpClientChannel` at all, and if it did there would be 5 channels passed even though there are only 4 unique sources of channel. I think you can simply drop the `tcpAcceptedChannel` altogether.

Reply:
you're right, thanks!